apps/rush-lib/src/api/RushConfiguration.ts

```diff
@@ -498,6 +498,8 @@ export class RushConfiguration {
   private _commonVersionsConfigurations: Map<string, CommonVersionsConfiguration> | undefined;
   // variant -> map of package name -> implicitly preferred version
   private _implicitlyPreferredVersions: Map<string, Map<string, string>> | undefined;
+  // variant -> map of package name to a set of specified dependency versions
+  private _allDependencyVersions: Map<string, Map<string, Set<string>>> | undefined;
```

Comment:
The data type `Map<string, Map<string, Set<string>>>` awkwardly combines four key pieces of information:
- The `variant: string` which must be a valid variant name
- The `packageName: string` which must follow the package name rules
- The `dependencyVersion: string` which must be a SemVer string
- The relationship between these things, i.e. a lookup table (variant, packageName) -> set of "all" versions

Nested types like `Map<string, Map<string, Set<string>>>` are bad because they anonymize this information.  Normally I would recommend to use some combination of "by" and plurals to combine that information into the identifier:

`versionSetsByPackageNameByVariant`

...then you can do:

```
const versionSetsByPackageName = versionSetsByPackageNameByVariant.get(variant);
const versionSet = versionSetsByPackageName.get(packageName);
if (versionSet.has(version)) { ...
```

...but with so much nesting, the above approach is still pretty clunky.  A better design is to ***declare your types*** rather than constructing them from algebra.  

Consider this small change:
```ts
interface IAnalysis {
  commonVersionsConfiguration: CommonVersionsConfiguration | undefined;
  implicitlyPreferredVersionByPackageName: Map<string, string> | undefined;
  allVersionsByPackageName: Map<string, Set<string>> | undefined;
}

private _analysisByVariant: Map<string, IAnalysis>;
```

It might even be more efficient, since we don't need to repeatedly look up the same `variant` for each part of the operation.

Reply:
(This engineering debt has been around since when variants were first introduced.  I'll defer to you as to whether adding another `Map<string, Map<string, Set<string>>> | undefined` is enough motivation to fix it in this PR.)

Reply:
I'll raise a separate PR to clean this up.